// @ngx-loading-bar/core/esm5/loading-bar.state.js
import { __assign } from "tslib";
import { Subject, timer, of } from 'rxjs';
import { map, switchMap, take, tap, startWith, shareReplay } from 'rxjs/operators';
var LoadingBarState = /** @class */ (function () {
    function LoadingBarState(config) {
        var _this = this;
        if (config === void 0) { config = {}; }
        this.config = config;
        this.state = {
            action: null,
            value: 0,
            initialValue: 0,
        };
        this.requests = null;
        this.disabled = false;
        this.stream$ = new Subject();
        this._value$ = null;
        this.timer$ = function (s) {
            var state$ = of(s);
            switch (s.action) {
                case 'start':
                case 'increment':
                case 'set': {
                    if (s.action === 'start' && _this.config.latencyThreshold === 0 && s.value === 0) {
                        s.value = s.initialValue;
                    }
                    if (_this.requests > 0) {
                        state$ = timer(_this.config.latencyThreshold, 250).pipe(map(function (t) { return (__assign(__assign({}, s), { value: t === 0 ? _this.state.value || s.initialValue : _this._increment() })); }));
                    }
                    break;
                }
                case 'complete':
                case 'stop': {
                    // Attempt to aggregate any start/complete calls within 500ms:
                    state$ =
                        s.value === 0
                            ? of(__assign({}, s))
                            : timer(0, 500).pipe(take(2), map(function (t) { return ({ value: t === 0 ? 100 : 0 }); }));
                    break;
                }
            }
            return state$.pipe(map(function (next) { return (__assign(__assign({}, next), { action: 'set' })); }), tap(function (next) { return _this.next(next, false); }));
        };
        this.config = __assign({ latencyThreshold: 0 }, config);
    }
    Object.defineProperty(LoadingBarState.prototype, "value$", {
        get: function () {
            var _this = this;
            if (this._value$) {
                return this._value$;
            }
            return (this._value$ = this.stream$.asObservable().pipe(startWith(this.state), switchMap(function (s) { return _this.timer$(s); }), shareReplay(), map(function (s) { return s.value; })));
        },
        enumerable: true,
        configurable: true
    });
    LoadingBarState.prototype.start = function (initialValue) {
        if (initialValue === void 0) { initialValue = 2; }
        if (this.disabled) {
            return;
        }
        this.next({ action: 'start', initialValue: initialValue });
    };
    LoadingBarState.prototype.stop = function () {
        this.next({ action: 'stop' });
    };
    LoadingBarState.prototype.complete = function () {
        this.next({ action: 'complete' });
    };
    LoadingBarState.prototype.disable = function () {
        this.disabled = true;
    };
    LoadingBarState.prototype.set = function (value) {
        this.next({ action: 'set', value: value });
    };
    LoadingBarState.prototype.increment = function (value) {
        if (value === void 0) { value = 0; }
        this.next({ action: 'increment', value: value });
    };
    LoadingBarState.prototype.next = function (state, emitEvent) {
        if (emitEvent === void 0) { emitEvent = true; }
        switch (state.action) {
            case 'start':
                this.requests = (this.requests || 0) + 1;
                break;
            case 'complete':
                this.requests = (this.requests || 1) - 1;
                if (this.requests > 0) {
                    return;
                }
                break;
            case 'stop':
                this.requests = 0;
                break;
            case 'increment':
                state.value = this._increment(state.value);
                break;
        }
        this.state = __assign(__assign(__assign({}, this.state), { action: null }), state);
        if (emitEvent) {
            this.stream$.next(this.state);
        }
    };
    LoadingBarState.prototype._increment = function (rnd) {
        if (rnd === void 0) { rnd = 0; }
        var stat = this.state.value;
        if (stat >= 99) {
            rnd = 0;
        }
        if (rnd === 0) {
            if (stat >= 0 && stat < 25) {
                // Start out between 3 - 6% increments
                rnd = Math.random() * (5 - 3 + 1) + 3;
            }
            else if (stat >= 25 && stat < 65) {
                // increment between 0 - 3%
                rnd = Math.random() * 3;
            }
            else if (stat >= 65 && stat < 90) {
                // increment between 0 - 2%
                rnd = Math.random() * 2;
            }
            else if (stat >= 90 && stat < 99) {
                // finally, increment it .5 %
                rnd = 0.5;
            }
            else {
                // after 99%, don't increment:
                rnd = 0;
            }
        }
        return rnd + stat;
    };
    return LoadingBarState;
}());
export { LoadingBarState };
