// angular-calendar/esm5/modules/common/calendar-drag-helper.provider.js
import { isInsideLeftAndRight, isWithinThreshold } from './util';
var CalendarDragHelper = /** @class */ (function () {
    function CalendarDragHelper(dragContainerElement, draggableElement) {
        this.dragContainerElement = dragContainerElement;
        this.startPosition = draggableElement.getBoundingClientRect();
    }
    CalendarDragHelper.prototype.validateDrag = function (_a) {
        var x = _a.x, y = _a.y, snapDraggedEvents = _a.snapDraggedEvents, dragAlreadyMoved = _a.dragAlreadyMoved, transform = _a.transform;
        var isDraggedWithinThreshold = isWithinThreshold({ x: x, y: y }) || dragAlreadyMoved;
        if (snapDraggedEvents) {
            var inner = Object.assign({}, this.startPosition, {
                left: this.startPosition.left + transform.x,
                right: this.startPosition.right + transform.x,
                top: this.startPosition.top + transform.y,
                bottom: this.startPosition.bottom + transform.y,
            });
            if (isDraggedWithinThreshold) {
                var outer = this.dragContainerElement.getBoundingClientRect();
                var isTopInside = outer.top < inner.top && inner.top < outer.bottom;
                var isBottomInside = outer.top < inner.bottom && inner.bottom < outer.bottom;
                return (isInsideLeftAndRight(outer, inner) && (isTopInside || isBottomInside));
            }
            /* istanbul ignore next */
            return false;
        }
        else {
            return isDraggedWithinThreshold;
        }
    };
    return CalendarDragHelper;
}());
export { CalendarDragHelper };
